lib/libc/gen/fts.3
.\" Copyright (c) 1989 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" %sccs.include.redist.man%
.\"
.\"	@(#)fts.3	5.6 (Berkeley) 6/23/90
.\"
.TH FTS 3 ""
.UC 7
.SH NAME
fts \- traverse a file hierarchy
.SH SYNOPSIS
.nf
.ft B
#include <sys/types.h>
#include <sys/stat.h>
#include <fts.h>

FTS *
ftsopen(path_argv, options, compar)
char *path_argv[];
int options, (*compar)();

FTSENT *
ftsread(ftsp);
FTS *ftsp;

FTSENT *
ftschildren(ftsp)
FTS *ftsp;

ftsset(ftsp, f, options)
FTS *ftsp;
FTSENT *f;
int options;

ftsclose(ftsp)
FTS *ftsp;
.ft R
.fi
.SH DESCRIPTION
The
.IR fts (3)
utilities are provided for traversing UNIX file hierarchies.
Two structures are defined (and typedef'd) in the include file <\fIfts.h\fP>.
The first is FTS, the structure that defines the file hierarchy stream.
The second is FTSENT, the structure that defines a file in the file
hierarchy.
.PP
The
.I ftsopen
routine takes a pointer to an array of character pointers (``argv'')
naming the file hierarchies to be traversed.
The array must be terminated by a pointer to a NULL string.
.PP
The 
.I options
specified are formed by
.IR or 'ing
one or more of the following values:
.TP
FTS_LOGICAL
This option causes 
.I ftsread
to use the function
.IR stat (2),
by default, to determine the status of each file.
If this option is set, the only symbolic links returned to the application
are those referencing non-existent files.
Either FTS_LOGICAL or FTS_PHYSICAL
.B must
be provided to the 
.I ftsopen
routine.
.TP
FTS_NOCHDIR
As a performance optimization,
.I ftsread
changes directories as it descends the hierarchy.
This has the side-effect that applications cannot rely on being
in any particular directory.
The FTS_NOCHDIR option turns off this optimization.
Note that applications should not change the current directory
(even if FTS_NOCHDIR is specified) unless absolute pathnames were
provided as arguments to
.IR ftsopen .
.TP
FTS_NOSTAT
By default,
.I ftsread
and
.I ftschildren
provide file characteristic information (the
.I statb
field) for each file they return.
This option relaxes that requirement; the contents of the
.I statb
field may be undefined, and the
.I info
field may be set to FTS_NS.
.TP
FTS_PHYSICAL
This option causes 
.I ftsread
to use the function
.IR lstat (2),
by default, to determine the status of each file.
If this option is set, all symbolic links are returned to the application
program.
Either FTS_LOGICAL or FTS_PHYSICAL
.B must
be provided to the 
.I ftsopen
routine.
.TP
FTS_SEEDOT
This option causes the routine
.I ftsread
to return structures for the directory entries ``.'' and ``..''.
By default they are ignored unless specified as an argument to
.IR ftsopen .
.TP
FTS_XDEV
This option keeps
.I fts
from descending into directories that have a different device number
than the file the descent began from.
.PP
The argument
.I compar
specifies a user-defined routine which is used to order the traversal
of directories.
.I Compar
takes two pointers to pointers to FTSENT structures as arguments and
should return a negative value, zero, or a positive value to indicate
if the file referenced by its first argument comes before, in any order
with respect to, or after, the file referenced by its second argument.
.PP
The
.I fts_accpath
and
.I fts_path
fields of the FTSENT structures may not be used in this comparison.
If the option
.I FTS_NOSTAT
is specified, the
.I fts_stab
field may not be used as well.
If the
.I compar
argument is NULL the directory traversal order is unspecified except
for the root paths which are traversed in the order listed in
.IR path_argv .
.PP
The
.I ftsclose
routine closes a file hierarchy stream and changes the current
directory to the directory
.I ftsopen
was called from.
.I Ftsclose
returns 0 on success, and -1 if an error occurs.
.PP
Each call to the
.I ftsread 
routine returns a pointer to an FTSENT structure describing a file in
the hierarchy.
Directories (that are readable, searchable and do not cause cycles)
are visited at least twice, before any of their descendants have been
visited (pre-order) and after all their descendants have been visited
(post-order).
All other files are visited at least once.
.PP
The FTSENT structure contains at least the following fields:
.sp
.RS
.nf
.ta .5i +.5i +\w'struct ftsent *parent;\0\0\0'u
typedef struct ftsent {
	struct ftsent *fts_parent;		/* parent structure */
	struct ftsent *fts_link;		/* cycle or file structure */
	union {
		long number;		/* local numeric value */
		void *pointer;		/* local address value */
	} fts_local;
	char *fts_accpath;			/* path from current directory */
	char *fts_path;			/* path from starting directory */
	char *fts_name;			/* file name */
	short fts_pathlen;			/* strlen(path) */
	short fts_namelen;			/* strlen(name) */
	short fts_level;			/* depth (\-1 to N) */
	unsigned short fts_info;		/* flags for entry */
	struct stat fts_statb;			/* stat(2) information */
} FTSENT;
.fi
.RE
.PP
The fields are as follows:
.TP
fts_parent
A pointer to a structure referencing the file in the hierarchy
immediately above the current file/structure.
A parent structure for the initial entry point is provided as well,
however, only the
.I fts_local
and
.I fts_level
fields are guaranteed to be initialized.
.TP
fts_link
If a directory causes a cycle in the hierarchy, either because of a
hard link between two directories, or a symbolic link pointing to a
directory, the
.I fts_link
field of the structure will point to the structure in the hierarchy 
that references the same file as the current structure.
Upon return from the
.I ftschildren
routine, the
.I fts_link
field points to the next structure in the linked list of entries
from the directory.
Otherwise, the contents of the
.I fts_link
field are undefined.
.TP
fts_local
This field is provided for the use of the application program.
It can be used to store either a long value or an address.
.TP
fts_accpath
A path that can be used to access the file.
.TP
fts_path
The path for the file relative to the root of the traversal.
.TP
fts_name
The basename of the file.
.TP
fts_pathlen
The length of the string referenced by
.IR fts_path .
.TP
fts_namelen
The length of the string referenced by
.IR fts_name .
.TP 
fts_level
The depth of the traversal, numbered from \-1 to N.
The parent structure of the root of the traversal is numbered \-1, and
the structure for the root of the traversal is numbered 0.
.TP 
fts_info
Information describing the file.
With the exception of directories (FTS_D), all of these entries are
terminal, i.e. they will not be revisited, nor will their descendants
be visited (if they have not been visited already).
.RS
.TP
FTS_D
A directory being visited in pre-order.
.TP
FTS_DC
A directory that causes a cycle.
The 
.I fts_link
field of the structure will be filled in as well.
.TP
FTS_DEFAULT
Anything that's not one of the others.
.TP
FTS_DNR
A directory that cannot be read.
.TP
FTS_DNX
A directory that cannot be searched.
.TP
FTS_DOT
A file named ``.'' or ``..'' with a
.I fts_level
field not equal to zero, i.e. one not specified as an argument to
.IR ftsopen .
(See FTS_SEEDOT.)
.TP
FTS_DP
A directory that is being visited in post-order.
The contents of the structure will be unchanged from when the
directory was visited in pre-order.
.TP
FTS_ERR
An error indicator; the external variable
.I errno
contains an error number indicating the reason for the error.
.TP
FTS_F
A regular file.
.TP
FTS_NS
No
.IR stat (2)
information is available at this time (see FTS_NOSTAT); the
contents of the
.I fts_statb
field are undefined.
.TP
FTS_SL
A symbolic link.
.TP
FTS_SLNONE
A symbolic link with a non-existent target.
.RE
.TP
fts_statb
.IR Stat (2)
information for the file.
.PP
The
.I fts_accpath
and
.I fts_path
fields are guaranteed to be NULL-terminated
.B only
for the file most recently returned by
.IR ftsread .
The
.I fts_name
field is always NULL-terminated.
To use these fields to reference any other active files may require
that they be modified using the information contained in the
.I fts_pathlen
field.
Any such modifications should be undone before further calls to
.I ftsread
are attempted.
.PP
If all the members of the hierarchy have been returned,
.I ftsread
returns NULL and sets the external variable
.I errno
to 0.
If an error unrelated to a file in the hierarchy occurs,
.I ftsread
returns NULL and sets errno appropriately.
Otherwise, a pointer to an FTSENT structure is returned, and an
error may or may not have occurred; see FTS_ERR above.
.PP
When the most recently returned file from 
.I ftsread
is a directory being visited in pre-order, 
.I ftschildren
returns a pointer to the first entry in a linked list (sorted by
the comparison routine, if provided) of the directory entries
it contains.
The linked list is linked through the
.I fts_link
field of the FTSENT structure.
Each call to
.I ftschildren
recreates the list.
Note, cycles are not detected until a directory is actually visited,
therefore
.I ftschildren
will never return a structure with an
.I fts_info
field set to FTS_DC.
If the current file is not a directory being visited in pre-order,
or if an error occurs, or the file does not contain any entries
.I ftschildren
returns NULL.
If an error occurs,
.I ftschildren
sets errno appropriately, otherwise it sets errno to zero.
.PP
The pointers returned by
.I ftsread
and
.I ftschildren
point to structures which may be overwritten.
Structures returned by
.I ftschildren
and
.I ftsread
may be overwritten after a call to
.I ftsclose
on the same file hierarchy.
Otherwise, structures returned by
.I ftschildren
will not be overwritten until a subsequent call to either
.I ftschildren
or
.I ftsread
on the same file hierarchy.
Structures returned by
.I ftsread
will not not be overwritten until a subsequent call to
.I ftsread
on the same file hierarchy stream, unless it represents a file of type
directory, in which case it will not be overwritten until after a
subsequent call to
.I ftsread
after it has been returned by the function
.I ftsread
in post-order.
.PP
The routine
.I ftsset
allows the user application to determine further processing for the
file
.I f
of the stream
.IR ftsp .
.I Ftsset
returns 0 on success, and -1 if an error occurs.
.I Option
may be set to any one of the following values.
.TP
FTS_AGAIN
Re-visit the file; any file type may be re-visited.
The next call to
.I ftsread
will return the referenced file.
The 
.I fts_stat
and
.I fts_info
fields of the structure will be reinitialized at that time,
but no other fields will have been modified.
This option is meaningful only for the most recently returned
file from
.IR ftsread .
Normal use is for post-order directory visits, where it causes the
directory to be re-visited (in both pre and post-order) as well as all
of its descendants.
.TP
FTS_FOLLOW
The referenced file must be a symbolic link.
If the referenced file is the one most recently returned by
.IR ftsread ,
the next call to
.I ftsread
returns the file with the
.I fts_info
and
.I fts_statb
fields reinitialized to reflect the target of the symbolic link instead
of the symbolic link itself.
If the file is one of those most recently returned by
.IR ftschildren ,
the
.I fts_info
and
.I fts_statb
fields of the structure, when returned by
.IR ftsread ,
will reflect the target of the symbolic link instead of the symbolic link
itself.
In either case, if the target of the link is a directory, the pre-order
return, followed by the return of all of its descendants, followed by a
post-order return, is done.
.TP
FTS_SKIP
No descendants of this file are visited.
.PP
Hard links between directories that do not cause cycles or symbolic
links to symbolic links may cause files to be visited more than once,
or directories more than twice.
.SH ERRORS
.I Ftsopen
may fail and set errno for any of the errors specified for the library
routine
.IR malloc (3).
.PP
.I Ftsclose
may fail and set errno for any of the errors specified for the library
routine
.IR chdir (2).
.PP
.I Ftsread
and
.I ftschildren
may fail and set errno for any of the errors specified for the library
routines
.IR chdir (2),
.IR getgroups (2),
.IR malloc (3),
.IR opendir (3),
.IR readdir (3)
and
.IR stat (2).
.SH SEE ALSO
find(1), chdir(2), stat(2), qsort(3)
.SH STANDARDS
The
.I fts
utility is expected to be a superset of the POSIX 1003.1 specification.


lib/libc/gen/fts.c
/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * %sccs.include.redist.c%
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)fts.c	5.5 (Berkeley) 5/22/90";
#endif /* LIBC_SCCS and not lint */

#include <sys/param.h>
#include <sys/stat.h>
#include <dirent.h>
#include <errno.h>
#include <fts.h>
#include <string.h>

extern int errno;

FTSENT *fts_alloc(), *fts_build(), *fts_cycle(), *fts_sort(), *fts_root();
short fts_stat();
char *malloc(), *realloc();

/*
 * Special case a root of "/" so that slashes aren't appended causing
 * paths to be written as "//foo".
 */
#define	NAPPEND(p) \
	(p->fts_level == ROOTLEVEL && p->fts_pathlen == 1 && \
	    p->fts_path[0] == '/' ? 0 : p->fts_pathlen)

#define	CHDIR(sp, path)	(!(sp->fts_options & FTS_NOCHDIR) && chdir(path))
#define	FCHDIR(sp, fd)	(!(sp->fts_options & FTS_NOCHDIR) && fchdir(fd))

#define	ROOTLEVEL	0
#define	ROOTPARENTLEVEL	-1

static FTS *stream;			/* current stream pointer */

FTS *
ftsopen(argv, options, compar)
	char *argv[];
	register int options;
	int (*compar)();
{
	register FTS *sp;
	register FTSENT *p, *root;
	register int nitems, maxlen;
	FTSENT *parent, *tmp;
	char wd[MAXPATHLEN], *getwd(), *strdup();

	/* allocate/initialize the stream */
	if (!(stream = sp = (FTS *)malloc((u_int)sizeof(FTS))))
		return(NULL);
	bzero(sp, sizeof(FTS));
	sp->fts_compar = compar;
	sp->fts_options = options;

	/* allocate/initialize root's parent */
	if (!(parent = fts_alloc("", 0)))
		goto mem1;
	parent->fts_level = ROOTPARENTLEVEL;

	/* allocate/initialize root(s) */
	if (options & FTS_MULTIPLE) {
		maxlen = -1;
		for (root = NULL, nitems = 0; *argv; ++argv, ++nitems) {
			if (!(p = fts_root(*argv)))
				goto mem2;
			if (maxlen < p->fts_namelen)
				maxlen = p->fts_namelen;
			/*
			 * if comparison routine supplied, traverse in sorted
			 * order; otherwise traverse in the order specified.
			 */
			if (compar) {
				p->fts_link = root;
				root = p;
				p->fts_accpath = p->fts_name;
				p->fts_info = fts_stat(p, 0);
			} else {
				p->fts_link = NULL;
				if (!root)
					tmp = root = p;
				else {
					tmp->fts_link = p;
					tmp = p;
				}
			}
			p->fts_level = ROOTLEVEL;
			p->fts_parent = parent;
		}
		if (compar && nitems > 1)
			root = fts_sort(root, nitems);
	} else {
		if (!(root = fts_root((char *)argv)))
			goto mem2;
		maxlen = root->fts_namelen;
		root->fts_link = NULL;
		root->fts_level = ROOTLEVEL;
		root->fts_parent = parent;
	}

	/* start out with at least 1K+ of path space */
	if (!fts_path(MAX(maxlen, MAXPATHLEN)))
		goto mem2;

	/*
	 * some minor trickiness.  Set the pointers so that ftsread thinks
	 * we've just finished the node before the root(s); set p->fts_info
	 * to FTS_NS so that everything about the "current" node is ignored.
	 * Rather than allocate a dummy node use the root's parent's link
	 * pointer.  This is handled specially in ftsclose() as well.
	 */
	sp->fts_cur = parent;
	parent->fts_link = root;
	parent->fts_info = FTS_NS;

	/*
	 * if using chdir(2), do a getwd() to insure that we can get back
	 * here; this could be avoided for some paths, but probably not
	 * worth the effort.  Slashes, symbolic links, and ".." are all
	 * fairly nasty problems.  Note, if we can't get the current
	 * working directory we run anyway, just more slowly.
	 */
	if (!(options & FTS_NOCHDIR) &&
	    (!getwd(wd) || !(sp->fts_wd = strdup(wd))))
		sp->fts_options |= FTS_NOCHDIR;

	return(sp);

mem2:	fts_lfree(root);
	(void)free((char *)parent);
mem1:	(void)free((char *)sp);
	return(NULL);
}

static
fts_load(p)
	register FTSENT *p;
{
	register int len;
	register char *cp;

	/*
	 * load the stream structure for the next traversal; set the
	 * accpath field specially so the chdir gets done to the right
	 * place and the user can access the first node.
	 */
	len = p->fts_pathlen = p->fts_namelen;
	bcopy(p->fts_name, stream->fts_path, len + 1);
	if ((cp = rindex(p->fts_name, '/')) && (cp != p->fts_name || cp[1])) {
		len = strlen(++cp);
		bcopy(cp, p->fts_name, len + 1);
		p->fts_namelen = len;
	}
	p->fts_accpath = p->fts_path = stream->fts_path;
}

ftsclose(sp)
	FTS *sp;
{
	register FTSENT *freep, *p;
	int saved_errno;

	if (sp->fts_cur)
		/* check for never having read anything */
		if (sp->fts_cur->fts_level == ROOTPARENTLEVEL)
			fts_lfree(sp->fts_cur);
		else {
			for (p = sp->fts_cur; p->fts_level > ROOTPARENTLEVEL;) {
				freep = p;
				p = p->fts_link ? p->fts_link : p->fts_parent;
				(void)free((char *)freep);
			}
			(void)free((char *)p);
		}

	/* free up child linked list, sort array, path buffer */
	if (sp->fts_child)
		fts_lfree(sp->fts_child);
	if (sp->fts_array)
		(void)free((char *)sp->fts_array);
	(void)free((char *)sp->fts_path);

	/*
	 * return to original directory, save errno if necessary;
	 * free up the directory buffer
	 */
	if (!(sp->fts_options & FTS_NOCHDIR)) {
		saved_errno = chdir(sp->fts_wd) ? errno : 0;
		(void)free(sp->fts_wd);
	}

	/* free up the stream pointer */
	(void)free((char *)sp);

	/* set errno and return */
	if (!(sp->fts_options & FTS_NOCHDIR) && saved_errno) {
		errno = saved_errno;
		return(-1);
	}
	return(0);
}

FTSENT *
ftsread(sp)
	register FTS *sp;
{
	register FTSENT *p;
	register int instr;
	static int cd;
	FTSENT *tmp;
	char *cp;

	/* if finished or unrecoverable error, return NULL */
	if (!sp->fts_cur || sp->fts_options & FTS__STOP)
		return(NULL);

	/* set global stream pointer, and current node pointer */
	stream = sp;
	p = sp->fts_cur;

	/* save and zero out user instructions */
	instr = p->fts_instr;
	p->fts_instr = 0;

	/* if used link pointer for cycle detection, restore it */
	if (sp->fts_savelink) {
		p->fts_link = sp->fts_savelink;
		sp->fts_savelink = NULL;
	}

	/* any type of file may be re-visited; re-stat and return */
	if (instr == FTS_AGAIN) {
		p->fts_info = fts_stat(p, 0);
		return(p);
	}

	if (p->fts_info == FTS_D)
		if (instr == FTS_SKIP || sp->fts_options & FTS_XDEV &&
		    p->fts_statb.st_dev != sp->sdev) {
			if (sp->fts_child) {
				fts_lfree(sp->fts_child);
				sp->fts_child = NULL;
			}
		} else {
			if (!sp->fts_child &&
			    (!(sp->fts_child = fts_build(sp, 1))))
				return(p);
			p = sp->fts_child;
			sp->fts_child = NULL;
			return(sp->fts_cur = p);
		}
	else if (p->fts_info == FTS_SL && instr == FTS_FOLLOW) {
		p->fts_info = fts_stat(p, 1);
		return(p);
	}

	/*
	 * user may have called ftsset on pointer returned by
	 * ftschildren; handle it here.
	 */
	for (p = p->fts_link; p;) {
		instr = p->fts_instr;
		if (instr == FTS_FOLLOW) {
			p->fts_info = fts_stat(p, 1);
			p->fts_instr = 0;
			break;
		}
		if (instr == FTS_SKIP) {
			tmp = p;
			p = p->fts_link;
			(void)free((char *)tmp);
			continue;
		}
		p->fts_instr = 0;
		break;
	}

	/* go to next node on this level */
	if (p) {
		/*
		 * if root level node, set up paths and return.  If not the
		 * first time, and it's not an absolute pathname, get back
		 * to wherever we started from.
		 */
		if (p->fts_level == ROOTLEVEL) {
			fts_load(p);
			if (cd) {
				(void)free((char *)sp->fts_cur);
				if (p->fts_path[0] != '/' &&
				    CHDIR(sp, sp->fts_wd)) {
					/* return target path for error msg */
					p->fts_path = sp->fts_wd;
					p->fts_info = FTS_ERR;
					sp->fts_options |= FTS__STOP;
					return(sp->fts_cur = p);
				}
			} else
				cd = 1;
			p->fts_info = fts_stat(p, 0);
			sp->sdev = p->fts_statb.st_dev;
		} else {
			(void)free((char *)sp->fts_cur);
			cp = sp->fts_path + NAPPEND(p->fts_parent);
			*cp++ = '/';
			bcopy(p->fts_name, cp, p->fts_namelen + 1);
			if (p->fts_info == FTS_D && (tmp = fts_cycle(p))) {
				sp->fts_savelink = p->fts_link;
				p->fts_link = tmp;
			}
		}
		return(sp->fts_cur = p);
	}

	/* go to parent */
	p = sp->fts_cur->fts_parent;
	(void)free((char *)sp->fts_cur);
	if (p->fts_level == ROOTPARENTLEVEL) {
		/*
		 * done; free everything up and set errno to 0 so the user
		 * can distinguish between error and EOF.
		 */
		(void)free((char *)p);
		errno = 0;
		return(sp->fts_cur = NULL);
	}

	sp->fts_path[p->fts_pathlen] = '\0';
	if (CHDIR(sp, "..")) {
		sp->fts_options |= FTS__STOP;
		p->fts_info = FTS_ERR;
	} else
		p->fts_info = FTS_DP;
	return(sp->fts_cur = p);
}

/*
 * ftsset takes the stream as an argument although it's not used in this
 * implementation; it would be necessary if anyone wanted to add global
 * semantics to fts using ftsset.  A possible error return is allowed for
 * similar reasons.
 */
/* ARGSUSED */
ftsset(sp, p, instr)
	FTS *sp;
	FTSENT *p;
	int instr;
{
	p->fts_instr = instr;
	return(0);
}

FTSENT *
ftschildren(sp)
	register FTS *sp;
{
	/*
	 * set errno to 0 so that user can tell the difference between an
	 * error and a directory without entries.
	 */
	errno = 0;
	if (sp->fts_cur->fts_info != FTS_D || sp->fts_options & FTS__STOP)
		return(NULL);
	if (sp->fts_child)
		fts_lfree(sp->fts_child);
	return(sp->fts_child = fts_build(sp, 0));
}

#define	ISDOT(a)	(a[0] == '.' && (!a[1] || a[1] == '.' && !a[2]))

static FTSENT *
fts_build(sp, set_node)
	register FTS *sp;
	int set_node;
{
	register struct dirent *dp;
	register FTSENT *p, *head;
	register int nitems;
	DIR *dirp;
	int descend, len, level, maxlen, nlinks, saved_errno;
	char *cp;

	p = sp->fts_cur;
	if (!(dirp = opendir(p->fts_accpath))) {
		if (set_node) {
			errno = 0;
			p->fts_info = FTS_DNR;
		}
		return(NULL);
	}

	/*
	 * the real slowdown in walking the tree is the stat calls.  If
	 * FTS_NOSTAT is set and it's a physical walk (so that symbolic
	 * links can't be directories), fts assumes that the number of
	 * subdirectories in a node is equal to the number of links to
	 * the parent.  This allows stat calls to be skipped in any leaf
	 * directories and for any nodes after the directories in the
	 * parent node have been found.  This empirically cuts the stat
	 * calls by about 2/3.
	 */
	nlinks =
	    sp->fts_options & FTS_NOSTAT && sp->fts_options & FTS_PHYSICAL ?
	    p->fts_statb.st_nlink - (sp->fts_options & FTS_SEEDOT ? 0 : 2) : -1;

	if (nlinks || set_node) {
		if (FCHDIR(sp, dirfd(dirp))) {
			if (set_node) {
				errno = 0;
				p->fts_info = FTS_DNX;
			}
			return(NULL);
		}
		descend = 1;
	} else
		descend = 0;

	/* get max file name length that can be stored in current path */
	maxlen = sp->fts_pathlen - p->fts_pathlen - 1;

	cp = sp->fts_path + (len = NAPPEND(p));
	*cp++ = '/';

	level = p->fts_level + 1;

	/* read the directory, attching each new entry to the `link' pointer */
	for (head = NULL, nitems = 0; dp = readdir(dirp);) {
		if (ISDOT(dp->d_name) && !(sp->fts_options & FTS_SEEDOT))
			continue;

		if (!(p = fts_alloc(dp->d_name, dp->d_namlen))) {
			saved_errno = errno;
			goto mem1;
		}
		if (dp->d_namlen > maxlen) {
			if (!fts_path((int)dp->d_namlen)) {
				/* quit: this stream no longer has a path */
				sp->fts_options |= FTS__STOP;
				saved_errno = errno;
				(void)free((char *)p);
mem1:				fts_lfree(head);
				if (set_node)
					p->fts_info = FTS_ERR;
				if (descend && CHDIR(sp, "..")) {
					saved_errno = errno;
					sp->fts_options |= FTS__STOP;
				}
				errno = saved_errno;
				return(NULL);
			}
			maxlen = sp->fts_pathlen - sp->fts_cur->fts_pathlen - 1;
		}

		p->fts_pathlen = len + dp->d_namlen + 1;
		p->fts_accpath =
		    sp->fts_options & FTS_NOCHDIR ? p->fts_path : p->fts_name;

		p->fts_parent = sp->fts_cur;
		p->fts_level = level;

		if (nlinks) {
			/* make sure fts_stat has a filename to stat */
			if (sp->fts_options & FTS_NOCHDIR)
				bcopy(p->fts_name, cp, p->fts_namelen + 1);
			p->fts_info = fts_stat(p, 0);
			if (nlinks > 0 && (p->fts_info == FTS_D ||
			    p->fts_info == FTS_DNR || p->fts_info == FTS_DNX))
				--nlinks;
		} else
			p->fts_info = FTS_NS;

		p->fts_link = head;
		head = p;
		++nitems;
	}
	(void)closedir(dirp);

	if (descend && (!nitems || !set_node) && CHDIR(sp, "..")) {
		sp->fts_options |= FTS__STOP;
		p->fts_info = FTS_ERR;
		*--cp = '\0';
		return(NULL);
	}

	if (!nitems) {
		if (set_node)
			p->fts_info = FTS_DP;
		*--cp = '\0';
		return(NULL);
	}

	if (sp->fts_compar && nitems > 1)
		head = fts_sort(head, nitems);

	if (set_node)
		bcopy(head->fts_name, cp, head->fts_namelen + 1);
	else
		*--cp = '\0';
	return(head);
}

static short
fts_stat(p, symflag)
	register FTSENT *p;
	int symflag;
{
	/*
	 * detection of symbolic links w/o targets.  If FTS_FOLLOW is set,
	 * the symlink structure is overwritten with the stat structure of
	 * the target.
	 */
	if (stream->fts_options & FTS_LOGICAL || symflag) {
		if (stat(p->fts_accpath, &p->fts_statb))
			return(symflag || !lstat(p->fts_accpath,
			    &p->fts_statb) ? FTS_SLNONE : FTS_ERR);
	} else if (lstat(p->fts_accpath, &p->fts_statb))
		return(FTS_ERR);

	switch(p->fts_statb.st_mode&S_IFMT) {
	case S_IFDIR:
		return(FTS_D);
	case S_IFLNK:
		return(FTS_SL);
	case S_IFREG:
		return(FTS_F);
	}
	return(FTS_DEFAULT);
}

static FTSENT *
fts_cycle(p)
	register FTSENT *p;
{
	register dev_t dev;
	register ino_t ino;

	/*
	 * cycle detection is brute force; if the tree gets deep enough or
	 * the number of symbolic links to directories is really high
	 * something faster might be worthwhile.
	 */
	dev = p->fts_statb.st_dev;
	ino = p->fts_statb.st_ino;
	for (p = p->fts_parent; p->fts_level > ROOTLEVEL; p = p->fts_parent)
		if (ino == p->fts_statb.st_ino && dev == p->fts_statb.st_dev)
			return(p);
	return(NULL);
}

#define	R(type, nelem, ptr) \
	(type *)realloc((char *)ptr, (u_int)((nelem) * sizeof(type)))

static FTSENT *
fts_sort(head, nitems)
	FTSENT *head;
	register int nitems;
{
	register FTSENT **ap, *p;

	/*
	 * construct an array of pointers to the structures and call qsort(3).
	 * Reassemble the array in the order returned by qsort.  If unable to
	 * sort for memory reasons, return the directory entries in their
	 * current order.  Allocate enough space for the current needs plus
	 * 40 so we don't realloc one entry at a time.
	 */
	if (nitems > stream->fts_nitems) {
		stream->fts_nitems = nitems + 40;
		if (!(stream->fts_array =
		    R(FTSENT *, stream->fts_nitems, stream->fts_array))) {
			stream->fts_nitems = 0;
			return(head);
		}
	}
	for (ap = stream->fts_array, p = head; p; p = p->fts_link)
		*ap++ = p;
	qsort((char *)stream->fts_array, nitems, sizeof(FTSENT *),
	    stream->fts_compar);
	for (head = *(ap = stream->fts_array); --nitems; ++ap)
		ap[0]->fts_link = ap[1];
	ap[0]->fts_link = NULL;
	return(head);
}

static FTSENT *
fts_alloc(name, len)
	char *name;
	register int len;
{
	register FTSENT *p;

	/*
	 * variable sized structures; the name is the last element so
	 * allocate enough extra space after the structure to hold it.
	 */
	if (!(p = (FTSENT *)malloc((u_int)(sizeof(FTSENT) + len))))
		return(NULL);
	bcopy(name, p->fts_name, len + 1);
	p->fts_namelen = len;
	p->fts_path = stream->fts_path;
	p->fts_instr = 0;
	p->fts_local.number = 0;
	p->fts_local.pointer = NULL;
	return(p);
}

static
fts_lfree(head)
	register FTSENT *head;
{
	register FTSENT *p;

	while (p = head) {
		head = head->fts_link;
		(void)free((char *)p);
	}
}

/*
 * allow essentially unlimited paths; certain programs (find, remove, ls)
 * need to work on any tree.  Most systems will allow creation of paths
 * much longer than MAXPATHLEN, even though the kernel won't resolve them.
 * Add an extra 128 bytes to the requested size so that we don't realloc
 * the path 2 bytes at a time.
 */
static
fts_path(size)
	int size;
{
	stream->fts_pathlen += size + 128;
	return((int)(stream->fts_path =
	    R(char, stream->fts_pathlen, stream->fts_path)));
}

static FTSENT *
fts_root(name)
	register char *name;
{
	register char *cp;

	/*
	 * rip trailing slashes; it's somewhat unclear in POSIX 1003.1 what
	 * /a/b/ really is, they don't talk about what a null path component
	 * resolves to.  This hopefully does what the user intended.  Don't
	 * allow null pathnames.
	 */
	for (cp = name; *cp; ++cp);
	if (cp == name) {
		errno = ENOENT;
		return(NULL);
	}
	while (--cp > name && *cp == '/');
	*++cp = '\0';
	return(fts_alloc(name, cp - name));
}
